jormungandr/src/utils/fire_forget_scheduler.rs

```diff
@@ -0,0 +1,286 @@
+use futures::{Async, Future, Poll, Stream};
+use std::collections::VecDeque;
+use std::convert::Infallible;
+use std::time::Duration;
+use thiserror::Error;
+use tokio::sync::mpsc::{
+    self,
+    error::{RecvError, TrySendError},
+    Receiver, Sender,
+};
+use tokio::timer::delay_queue::{DelayQueue, Key};
+
+#[derive(Error, Debug)]
+pub enum Error {
+    #[error("failed to send a command: {0}")]
+    CommandSend(&'static str),
+    #[error("failed to receive a command")]
+    CommandReceive(#[from] RecvError),
+    #[error("command queue closed")]
+    CommandQueueClosed,
+    #[error("timer error")]
+    Timer(#[from] tokio::timer::Error),
+}
+
+impl<T> From<TrySendError<T>> for Error {
+    fn from(error: TrySendError<T>) -> Self {
+        let cause = if error.is_closed() {
+            "channel closed"
+        } else if error.is_full() {
+            "no available capacity"
+        } else {
+            "unknown channel error"
+        };
+        Error::CommandSend(cause)
+    }
+}
+
+/// Schedule for fire-forget tasks
+///
+/// Each task has an ID (TID) and can be launched onto a worker identified with an ID (WID).
```

Comment:
Why do we need to concern ourselves with explicit worker IDs in this API?
It would be much simpler to have a single-dimensional table keyed with the target block ID, with a limited list (or `FuturesUnordered`) of running task objects in each entry, and maybe a list (no matter the order here, could be a `Vec` with LIFO) of pending pull states for tasks that will be launched when running tasks fail.

Reply:
The only reason WID is exposed is to de-duplicate requests and not allow flooding the same (possibly broken) node with copies of the same request while starving others. If that's not a concern, WID is useless and NodeId can be part of Data.

Reply:
Right. I still think these abstractions do not apply well to the subject matter, and we should not try to create this as a `utils` thing. Let's try with a purpose-specific `ConcurrentFetch` or something like that, and get rid of the generics where not needed: we only use the block hash for the grouping key (not "ID"), and the node ID as a discriminant for competing fetch jobs.

Reply:
Should we assume that all `ConcurrentFetch` jobs are always going to be identifiable by a `HeaderHash`-`NodeId` pair? It's easy to specialize `FireForgetSchedule` into `ConcurrentFetch` in current setup with a single line `type`, but the other way is irreversible.

Reply:
There may be custom logic involved. For example, we may choose not to start a new pull with a node where a slow pull for an earlier block ID is still in progress, but update the target block ID for the next pending pull job.